src/lib/comms/sol-mqtt.c

```diff
@@ -0,0 +1,579 @@
+            SOL_WRN("Couldn't handle mqtt handler that has unsupported " \
+                "version '%u', expected version is '%u'", \
+                ptr->api_version, SOL_MQTT_CONFIG_API_VERSION); \
+            return __VA_ARGS__; \
+        } \
+    } while (0)
+
+struct sol_mqtt {
+    struct mosquitto *mosq;
+
+    struct sol_fd *socket_read;
+    struct sol_fd *socket_write;
+
+    struct sol_timeout *connect_timeout;
+    struct sol_timeout *disconnect_timeout;
+    struct sol_timeout *publish_timeout;
+    struct sol_timeout *subscribe_timeout;
+    struct sol_timeout *unsubscribe_timeout;
+    struct sol_timeout *message_timeout;
+
+    const struct sol_mqtt_config *config;
+
+    void *data;
+
+    int socket_fd;
+
+    int connection_status;
+
+    time_t keepalive;
+};
+
+struct sol_mqtt_message {
+    char *topic;
+    void *payload;
+    size_t payload_len;
+    int id;
+    sol_mqtt_qos qos;
+    bool retain;
+};
+
+static void
+sol_mqtt_init(void)
+{
+    SOL_LOG_INTERNAL_INIT_ONCE;
+    init_ref++;
+    mosquitto_lib_init();
+}
+
+static void
+sol_mqtt_shutdown(void)
+{
+    init_ref--;
+
+    if (init_ref == 0)
+        mosquitto_lib_cleanup();
+}
+
+static bool
+sol_mqtt_event_loop(void *data, int fd, unsigned int active_flags)
+{
+    struct sol_mqtt *mqtt = data;
+    int r;
+
+    r = mosquitto_loop_read(mqtt->mosq, 1);
+    r |= mosquitto_loop_write(mqtt->mosq, 1);
+    r |= mosquitto_loop_misc(mqtt->mosq);
+
+    if (r != 0) {
+        SOL_WRN("Unable to perform mqtt socket operation");
+        goto remove;
+    }
+
+    /* No more data to write, remove SOL_FD_FLAGS_OUT event watcher */
+    if (active_flags & SOL_FD_FLAGS_OUT && !mosquitto_want_write(mqtt->mosq))
+        goto remove;
+
+    return true;
+
+remove:
+    if (active_flags & SOL_FD_FLAGS_OUT)
+        mqtt->socket_write = NULL;
+    else
+        mqtt->socket_read = NULL;
+
+    return false;
+}
+
+/*
+ * When mosquitto calls the a user provided callback, it's internal lock is held
+ * so we need to go back to solleta mainloop before calling our user callback
+ * in order to prevent a deadlock.
+ */
+
+#define CALLBACK_WRAPPER(_cb) \
+    static bool \
+    sol_mqtt_on_ ## _cb ## _wrapper(void * data) \
+    { \
+        struct sol_mqtt *mqtt = data; \
+        SOL_NULL_CHECK(mqtt, false); \
+        mqtt->_cb ## _timeout = NULL; \
+        if (mqtt->config->_cb) \
+            mqtt->config->_cb(mqtt, mqtt->data); \
+        return false; \
+    }
+
+CALLBACK_WRAPPER(connect);
+CALLBACK_WRAPPER(disconnect);
+CALLBACK_WRAPPER(publish);
+CALLBACK_WRAPPER(subscribe);
+CALLBACK_WRAPPER(unsubscribe);
+
+static void
+sol_mqtt_on_connect(struct mosquitto *mosq, void *data, int rc)
+{
+    struct sol_mqtt *mqtt = data;
+
+    SOL_NULL_CHECK(mqtt);
+
+    mqtt->connection_status = rc;
+
+    if (mqtt->connect_timeout || !mqtt->config || !mqtt->config->connect)
+        return;
+
+    mqtt->connect_timeout = sol_timeout_add(0, sol_mqtt_on_connect_wrapper, mqtt);
```

Comment:
As we talked irl, this is like this because mosquitto is not reentrant, so we need to let it return to release the internal lock. There is a comment about this a couple lines above.